src/include/serial_inline.h

```diff
@@ -126,9 +126,10 @@ __col_append_serial_func(WT_SESSION_IMPL *session, WT_INSERT_HEAD *ins_head, WT_
      * Set the calling cursor's record number. If we extended the file, update the last record
      * number.
      */
-    *recnop = recno;
-    if (recno > btree->last_recno)
-        btree->last_recno = recno;
+    do {
+        *recnop = recno;
+    } while (!__wt_atomic_cas64(&btree->last_recno, btree->last_recno, recno) &&
+      (recno > btree->last_recno));
```

Comment:
A couple things here. Firstly I think the conditional needs to be the other way around, since this code will still attempt the CAS if `recno` is less than `btree->last_recno`, which isn't the desired behaviour.

Secondly, if we're using `btree->last_recno` twice, we need to read it once and store it. If we swap the order of the conditional, i.e. write `while ((recno > btree->last_recno) && !__wt_atomic_cas64(&btree->last_recno, btree->last_recno, recno))`, there's a chance that `btree->last_recno` could get bumped between these two checks, our `recno` is now less than `btree->last_recno`, but the swap succeeds anyway because we re-read it.

Finally, do we expect/assume an atomic CAS to have both acquire and release semantics across all platforms? I know Arm does, not sure about any others. The reason I ask is to prevent any sort of read/write reordering across the CAS. I think this is one for @luke-pearson.

Reply:
Right now CAS is SEQ_CST which is a stronger guarantee than Acq/Rel but similar either way, nothing is getting reordered across it.

Reply:
Great - thanks. I believe it's just sufficient to make the first two fixes I suggested then.

Reply:
Thanks for the comment, updated here [235b0bc](https://github.com/wiredtiger/wiredtiger/pull/9656/commits/235b0bcb3e63b989e755d37104310a77ccdbf6ce)